pkg/alertmanager/distributor.go

```diff
@@ -0,0 +1,235 @@
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/httpgrpc/server"
+	"github.com/weaveworks/common/user"
+
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/ring/client"
+	"github.com/cortexproject/cortex/pkg/tenant"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+// Distributor forwards requests to individual alertmanagers.
+type Distributor struct {
+	services.Service
+
+	cfg              *MultitenantAlertmanagerConfig
+	requestsInFlight sync.WaitGroup
+
+	alertmanagerRing        ring.ReadRing
+	alertmanagerClientsPool ClientsPool
+
+	logger log.Logger
+}
+
+// NewDistributor constructs a new Distributor
+func NewDistributor(cfg *MultitenantAlertmanagerConfig, alertmanagersRing *ring.Ring, alertmanagerClientsPool ClientsPool, logger log.Logger, reg prometheus.Registerer) (d *Distributor, err error) {
+	if alertmanagerClientsPool == nil {
+		alertmanagerClientsPool = newAlertmanagerClientsPool(client.NewRingServiceDiscovery(alertmanagersRing), cfg.AlertmanagerClient, logger, reg)
+	}
+
+	d = &Distributor{
+		cfg:                     cfg,
+		logger:                  logger,
+		alertmanagerRing:        alertmanagersRing,
+		alertmanagerClientsPool: alertmanagerClientsPool,
+	}
+
+	d.Service = services.NewBasicService(nil, d.running, nil)
+	return d, nil
+}
+
+func (d *Distributor) running(ctx context.Context) error {
+	<-ctx.Done()
+	d.requestsInFlight.Wait()
+	return nil
+}
+
+// IsPathSupported returns true if the given route is currently supported by the Distributor.
+func (d *Distributor) IsPathSupported(path string) bool {
+	// API can be found at https://petstore.swagger.io/?url=https://raw.githubusercontent.com/prometheus/alertmanager/master/api/v2/openapi.yaml.
+	return strings.HasSuffix(path, "/alerts") ||
+		strings.HasSuffix(path, "/alerts/groups")
+}
+
+// DistributeRequest shards the writes and returns as soon as the quorum is satisfied.
+// In case of reads, it proxies the request to one of the alertmanagers.
+// DistributeRequest assumes that the caller has verified IsPathSupported return
+// true for the route.
+func (d *Distributor) DistributeRequest(w http.ResponseWriter, r *http.Request) {
+	d.requestsInFlight.Add(1)
+	defer d.requestsInFlight.Done()
+
+	userID, err := tenant.TenantID(r.Context())
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
+
+	logger := log.With(d.logger, "user", userID)
+
+	if r.Method == http.MethodGet || r.Method == http.MethodHead {
+		d.doRead(userID, w, r, logger)
+	} else {
+		d.doWrite(userID, w, r, logger)
+	}
+}
+
+func (d *Distributor) doWrite(userID string, w http.ResponseWriter, r *http.Request, logger log.Logger) {
+	var body []byte
+	var err error
+	if r.Body != nil {
+		body, err = ioutil.ReadAll(r.Body)
+		if err != nil {
+			if errors.Is(util.ErrRequestBodyTooLarge, err) {
```

Comment:
If we are setting a limit on http body, I would expect it to be for every request that comes to alertmanager and not special cased for some requests. And this is not the only place where we are reading the body, `server.HTTPRequest` also reads the body (which makes me realise we need this check there too). Also when alertmanager is processing the request without going through the distributor.

Reply:
Setting generic limit on http handler is wrong approach for several reasons:
- in general, large http request body is not a problem
- handlers that don’t read body **at all** don’t need the limit, nothing will break. Other handlers may need different limits (eg. if handler streams body to a file, 1MB limit may be too low).
- **it only becomes a problem when full body is read to memory** like here – we risk OOMs
- so **this** code needs to be protected – and that protection exists, but is hidden. But **this** code cannot check whether it is already using limited body or not, and blindly relies on the fact that it is. If we copy this handler code in the future to a different place, suddenly the protection may not be there.

For `server.HTTPRequest` I’d suggest giving it explicit max-body-size parameter – that breaks existing API, but in this case it may be good as it forces us to fix it everywhere.

Reply:
Made it local